nn/transformer.py

```diff
@@ -0,0 +1,231 @@
+      output = self.norm(output, epsilon=self.norm_eps)
+
+    return output
+
+
+class TransformerDecoderLayer(Module):
+  """
+  Defines one layer of a standard transformer decoder
+  """
+  def __init__(self, dim_model: int, num_heads: int, dim_ff: int = 2048, drop: float = 0.1, activation=relu,
+               layer_norm_eps: float = 1e-5, norm_first: bool = False):
+    """
+    :param dim_model: hidden dim, PyTorch name: d_model
+    :param num_heads: number heads, PyTorch name: nhead
+    :param dim_ff: dimension of feedforward layer, PyTorch name: dim_feedforward
+    :param drop: Dropout value, PyTorch name: dropout
+    :param activation: activation functional
+    :param layer_norm_eps: Epsilon value for layer normalization
+    :param norm_first: Whether to do layer norm before or afterwards
+    """
+    super().__init__()
+    self.dropout = drop
+    self.self_attn = MultiheadAttention(dim_model, num_heads, dropout=self.dropout)  # will change with AttentionModules
+    self.multihead_attn = MultiheadAttention(dim_model, num_heads, dropout=self.dropout)  # will change with AM
+
+    self.linear1 = Linear(dim_ff)
+    self.linear2 = Linear(dim_model)
+
+    self.norm_first = norm_first
+    self.norm_eps = layer_norm_eps
+    self.activationivation = activation
+
+  def forward(self, tgt: LayerRef, memory: LayerRef) -> LayerRef:
+    """
+    Two possible forward variants of decoder, defined by self.norm_first
+    """
+    x = tgt
+    if self.norm_first:
+      x = x + self._sa_block(layer_norm(x, epsilon=self.norm_eps))
+      x = x + self._mha_block(layer_norm(x, epsilon=self.norm_eps), memory)
+      x = x + self._ff_block(layer_norm(x, epsilon=self.norm_eps))
+    else:
+      x = layer_norm(x + self._sa_block(x), epsilon=self.norm_eps)
+      x = layer_norm(x + self._mha_block(x, memory), epsilon=self.norm_eps)
+      x = layer_norm(x + self._ff_block(x), epsilon=self.norm_eps)
+
+    return x
+
+  # self-attention block
+  def _sa_block(self, x: LayerRef) -> LayerRef:
+    x = self.self_attn(x, x, x)  # will change with Attention Modules
+    return dropout(x, self.dropout)
+
+  # multihead attention block
+  def _mha_block(self, x: LayerRef, mem: LayerRef) -> LayerRef:
+    x = self.multihead_attn(x, mem, mem)  # will change with Attention Modules
+    return dropout(x, self.dropout)
+
+  # feed forward block
+  def _ff_block(self, x: LayerRef) -> LayerRef:
+    x = self.linear2(dropout(self.activationivation(self.linear1(x)), self.dropout))
+    return dropout(x, self.dropout)
+
+
+class TransformerDecoder(Module):
+  """
+  Defines the full Decoder of the standard transformer
+  """
+  def __init__(self, decoder_layer: Union[TransformerDecoderLayer, Any], num_layers: int, norm=None,
+               layer_norm_eps: float = 1e-5):
+      """
+      :param decoder_layer: Decoder layer to be stacked num_layers times
+      :param num_layers: Number of layers
+      :param norm: normalization functional
```

Comment:
Add to the docstring that this is only the final output norm.

Reply:
Why do you call it "functional"? Call it just "function".